panic in system-tests: in Data-Node after 13 blocks

# Problem encountered
panic: failed to snapshot data:failed to copy table data:failed to execute copy 

# Where
System tests [RUN](https://jenkins.ops.vega.xyz/job/common/job/system-tests-wrapper/19082//)

# Release / Version
develop

# Details
Example log: https://jenkins.ops.vega.xyz/job/common/job/system-tests-wrapper/19082/artifact/testnet/logs/testnet-nodeset-full-2-full/data-node-2.stderr-2022-10-25T02%3A11%3A59Z.log

```
panic: failed to snapshot data:failed to copy table data:failed to execute copy copy (select * from last_block order by onerow_check) to '/snapshotsCopyTo/testnet-12-currentstatesnapshot/last_block': ERROR: could not open file "/snapshotsCopyTo/testnet-12-currentstatesnapshot/last_block" for writing: No such file or directory (SQLSTATE 58P01)

goroutine 1185 [running]:
code.vegaprotocol.io/vega/datanode/dehistory/snapshot.(*Service).CreateSnapshot.func5()
	/jenkins/workspace/common/system-tests-wrapper/vega/datanode/dehistory/snapshot/service_create_snapshot.go:87 +0x374
created by code.vegaprotocol.io/vega/datanode/dehistory/snapshot.(*Service).CreateSnapshot
	/jenkins/workspace/common/system-tests-wrapper/vega/datanode/dehistory/snapshot/service_create_snapshot.go:81 +0xaed
```

```
2022-10-25T02:22:16.879Z	INFO	root	start/node.go:143	Vega data node startup complete
2022-10-25T02:22:16.879Z	INFO	api.grpc	api/server.go:333	Starting gRPC based API	{"addr": "0.0.0.0", "port": "3027"}
2022-10-25T02:22:16.879Z	INFO	gateway.gql	graphql/server.go:147	Starting GraphQL based API	{"addr": "0.0.0.0", "port": 3028}
2022-10-25T02:22:16.880Z	WARN	gateway.gql	graphql/server.go:184	graphql playground enabled, this is not a recommended setting for production
2022-10-25T02:22:16.880Z	WARN	gateway.gql	graphql/server.go:244	GraphQL server is not configured to use HTTPS, which is required for subscriptions to work. Please see README.md for help configuring
2022-10-25T02:22:16.880Z	INFO	root	broker/socket_server.go:86	Starting broker socket server	{"addr": "0.0.0.0", "port": 3025}
2022-10-25T02:22:16.880Z	INFO	gateway.restproxy	rest/server.go:79	Starting REST<>GRPC based API	{"addr": "0.0.0.0", "port": 3029}
2022-10-25T02:22:20.116Z	INFO	root	broker/socket_server.go:90	New broker connection event	{"eventType": "Attaching", "id": 972335048, "address": "tcp://[::]:3025"}
2022-10-25T02:22:20.116Z	INFO	root	broker/socket_server.go:90	New broker connection event	{"eventType": "Attached", "id": 972335048, "address": "tcp://[::]:3025"}
2022-10-25T02:22:20.201Z	INFO	root	broker/sqlstore_broker.go:175	first unprocessed block received, starting block processing
2022-10-25T02:22:20.241Z	WARN	root	broker/sqlstore_broker.go:125	slow time update detected, time between checks 2s, block height: 1, total block processing time: 39.794745ms
2022-10-25T02:22:21.855Z	INFO	deHistory.snapshot	snapshot/service.go:44	creating snapshot for {History Snapshot for Chain ID:testnet Height From:1 Height To:12}
2022-10-25T02:22:21.960Z	INFO	deHistory.snapshot	snapshot/service_create_snapshot.go:83	copying all table data....
2022-10-25T02:22:21.964Z	ERROR	deHistory.snapshot	snapshot/service_create_snapshot.go:85	failed to snapshot data	{"error": "failed to copy table data:failed to execute copy copy (select * from last_block order by onerow_check) to '/snapshotsCopyTo/testnet-12-currentstatesnapshot/last_block': ERROR: could not open file \"/snapshotsCopyTo/testnet-12-currentstatesnapshot/last_block\" for writing: No such file or directory (SQLSTATE 58P01)"}
```

--- a/cmd/data-node/commands/dehistory/load.go
+++ b/cmd/data-node/commands/dehistory/load.go
@@ -2,6 +2,7 @@ package dehistory
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"code.vegaprotocol.io/vega/datanode/dehistory/initialise"
@@ -41,18 +42,27 @@ func (cmd *loadCmd) Execute(_ []string) error {
 		return fmt.Errorf("failed to read configuration:%w", err)
 	}
 
-	err = verifyChainID(log, cmd.SQLStore.ConnectionConfig, cmd.ChainID)
+	hasSchema, err := initialise.HasVegaSchema(context.Background(), cmd.SQLStore.ConnectionConfig)
 	if err != nil {
-		return fmt.Errorf("failed to verify chain id:%w", err)
+		return fmt.Errorf("failed to check for existing schema:%w", err)
+	}
+
+	if hasSchema {
+		err = verifyChainID(log, cmd.SQLStore.ConnectionConfig, cmd.ChainID)
+		if err != nil {
+			if !errors.Is(err, initialise.ErrChainNotFound) {
+				return fmt.Errorf("failed to verify chain id:%w", err)
+			}
+		}
 	}
 
 	if datanodeLive(cmd.Config) {
 		return fmt.Errorf("datanode must be shutdown before data can be loaded")
 	}
 
-	snapshotsCopyFrom, snapshotsCopyTo := initialise.GetSnapshotPaths(bool(cmd.Config.SQLStore.UseEmbedded), cmd.Config.DeHistory.Snapshot, vegaPaths)
-
-	snapshotService, err := snapshot.NewSnapshotService(log, cmd.Config.DeHistory.Snapshot, cmd.Config.SQLStore.ConnectionConfig, snapshotsCopyTo)
+	snapshotService, err := snapshot.NewSnapshotService(log, cmd.Config.DeHistory.Snapshot, cmd.Config.SQLStore.ConnectionConfig,
+		vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyFrom),
+		vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyTo))
 	if err != nil {
 		return fmt.Errorf("failed to create snapshot service: %w", err)
 	}
@@ -66,7 +76,9 @@ func (cmd *loadCmd) Execute(_ []string) error {
 	}
 
 	deHistoryService, err := dehistory.NewWithStore(ctx, log, cmd.Config.ChainID, cmd.Config.DeHistory,
-		cmd.Config.SQLStore.ConnectionConfig, snapshotService, deHistoryStore, cmd.Config.API.Port, snapshotsCopyFrom, snapshotsCopyTo)
+		cmd.Config.SQLStore.ConnectionConfig, snapshotService, deHistoryStore, cmd.Config.API.Port,
+		vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyFrom),
+		vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyTo))
 	if err != nil {
 		return fmt.Errorf("failed new dehistory service:%w", err)
 	}

--- a/cmd/data-node/commands/start/node_pre.go
+++ b/cmd/data-node/commands/start/node_pre.go
@@ -252,13 +252,6 @@ func (l *NodeCommand) preRun([]string) (err error) {
 }
 
 func (l *NodeCommand) initialiseDecentralizedHistory() error {
-	useEmbedded := bool(l.conf.SQLStore.UseEmbedded)
-	snapshotsCopyFromDir, snapshotsCopyToDir := initialise.GetSnapshotPaths(useEmbedded, l.conf.DeHistory.Snapshot, l.vegaPaths)
-	if useEmbedded {
-		l.conf.DeHistory.Snapshot.DatabaseSnapshotsCopyFromPath = snapshotsCopyFromDir
-		l.conf.DeHistory.Snapshot.DatabaseSnapshotsCopyToPath = snapshotsCopyToDir
-	}
-
 	deHistoryLog := l.Log.Named("deHistory")
 	deHistoryLog.SetLevel(l.conf.DeHistory.Level.Get())
 
@@ -266,13 +259,16 @@ func (l *NodeCommand) initialiseDecentralizedHistory() error {
 	deHistoryServiceLog := deHistoryLog.Named("service")
 
 	var err error
-	l.snapshotService, err = snapshot.NewSnapshotService(snapshotServiceLog, l.conf.DeHistory.Snapshot, l.conf.SQLStore.ConnectionConfig, snapshotsCopyToDir)
+	l.snapshotService, err = snapshot.NewSnapshotService(snapshotServiceLog, l.conf.DeHistory.Snapshot,
+		l.conf.SQLStore.ConnectionConfig, l.vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyFrom),
+		l.vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyTo))
 	if err != nil {
 		return fmt.Errorf("failed to create snapshot service:%w", err)
 	}
 
 	l.deHistoryService, err = dehistory.New(l.ctx, deHistoryServiceLog, l.conf.DeHistory, l.vegaPaths.StatePathFor(paths.DataNodeDeHistoryHome),
-		l.conf.SQLStore.ConnectionConfig, l.conf.ChainID, l.snapshotService, l.conf.API.Port, snapshotsCopyFromDir, snapshotsCopyToDir)
+		l.conf.SQLStore.ConnectionConfig, l.conf.ChainID, l.snapshotService, l.conf.API.Port, l.vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyFrom),
+		l.vegaPaths.StatePathFor(paths.DataNodeDeHistorySnapshotCopyTo))
 
 	if err != nil {
 		return fmt.Errorf("failed to create deHistory service:%w", err)

--- a/datanode/dehistory/config.go
+++ b/datanode/dehistory/config.go
@@ -29,7 +29,7 @@ type Config struct {
 func NewDefaultConfig() Config {
 	return Config{
 		Level:                encoding.LogLevel{Level: logging.InfoLevel},
-		Enabled:              false,
+		Enabled:              true,
 		WipeOnStartup:        true,
 		AddSnapshotsToStore:  true,
 		AddSnapshotsInterval: encoding.Duration{Duration: 5 * time.Second},

--- a/datanode/dehistory/initialise/initialise.go
+++ b/datanode/dehistory/initialise/initialise.go
@@ -6,14 +6,11 @@ import (
 	"fmt"
 	"math/rand"
 	"net"
-	"path/filepath"
 	"sort"
 	"strconv"
 	"time"
 
 	"code.vegaprotocol.io/vega/datanode/service"
-
-	"code.vegaprotocol.io/vega/paths"
 
 	"code.vegaprotocol.io/vega/datanode/dehistory/snapshot"
 	"code.vegaprotocol.io/vega/datanode/dehistory/store"
@@ -26,7 +23,10 @@ import (
 	"google.golang.org/grpc"
 )
 
-var ErrFailedToGetSegment = errors.New("no history segment found")
+var (
+	ErrFailedToGetSegment = errors.New("no history segment found")
+	ErrChainNotFound      = errors.New("no chain found")
+)
 
 type deHistoryService interface {
 	GetActivePeerAddresses() []string
@@ -304,19 +304,6 @@ func FetchHistoryBlocks(ctx context.Context, logInfo func(s string, args ...inte
 	return blocksFetched, nil
 }
 
-func GetSnapshotPaths(useEmbedded bool, conf snapshot.Config, vegaPaths paths.Paths) (snapshotsCopyFrom string, snapshotsCopyTo string) {
-	if useEmbedded {
-		embeddedPostgresRuntimeDir := vegaPaths.StatePathFor(paths.DataNodeEmbeddedPostgresRuntimeDir)
-		snapshotsCopyFrom = filepath.Join(embeddedPostgresRuntimeDir, conf.DatabaseSnapshotsCopyFromPath)
-		snapshotsCopyTo = filepath.Join(embeddedPostgresRuntimeDir, conf.DatabaseSnapshotsCopyToPath)
-	} else {
-		deHistoryHome := vegaPaths.StatePathFor(paths.DataNodeDeHistoryHome)
-		snapshotsCopyFrom = filepath.Join(deHistoryHome, conf.DatabaseSnapshotsCopyFromPath)
-		snapshotsCopyTo = filepath.Join(deHistoryHome, conf.DatabaseSnapshotsCopyToPath)
-	}
-	return snapshotsCopyFrom, snapshotsCopyTo
-}
-
 func VerifyChainID(chainID string, chainService *service.Chain) error {
 	if len(chainID) == 0 {
 		return errors.New("chain id must be set")
@@ -324,6 +311,10 @@ func VerifyChainID(chainID string, chainService *service.Chain) error {
 
 	currentChainID, err := chainService.GetChainID()
 	if err != nil {
+		if errors.Is(err, entities.ErrChainNotFound) {
+			return ErrChainNotFound
+		}
+
 		return fmt.Errorf("failed to get chain id:%w", err)
 	}
 

--- a/datanode/dehistory/service_test.go
+++ b/datanode/dehistory/service_test.go
@@ -537,11 +537,9 @@ func setupSnapshotService(testDbConfig sqlstore.Config, snapshotCopyFromPath str
 
 func setupSnapshotServiceWithNetworkParamFunc(testDbConfig sqlstore.Config, snapshotCopyFromPath string, snapshotCopyToPath string) *snapshot.Service {
 	snapshotServiceCfg := snapshot.NewDefaultConfig()
-	snapshotServiceCfg.DatabaseSnapshotsCopyToPath = snapshotCopyToPath
-	snapshotServiceCfg.DatabaseSnapshotsCopyFromPath = snapshotCopyFromPath
 
 	snapshotService, err := snapshot.NewSnapshotService(logging.NewTestLogger(), snapshotServiceCfg,
-		testDbConfig.ConnectionConfig, snapshotCopyToPath)
+		testDbConfig.ConnectionConfig, snapshotCopyFromPath, snapshotCopyToPath)
 	if err != nil {
 		panic(err)
 	}

--- a/datanode/dehistory/snapshot/config.go
+++ b/datanode/dehistory/snapshot/config.go
@@ -7,21 +7,15 @@ import (
 )
 
 type Config struct {
-	PanicOnSnapshotCreationError encoding.Bool `long:"panic-on-snapshot-creation-error" description:""`
-
-	DatabaseSnapshotsCopyToPath   string `long:"database-snapshot-copy-to-path" description:"the snapshots copy to path relative to the database working directory"`
-	DatabaseSnapshotsCopyFromPath string `long:"database-snapshot-copy-from-path" description:"the snapshots copy from path relative to the database working directory"`
-
-	WaitForCreationLockTimeout encoding.Duration `long:"wait-for-creation-lock-timeout" description:"the maximum a caller to create snapshot should have to wait to acquire the creation lock"`
+	PanicOnSnapshotCreationError encoding.Bool     `long:"panic-on-snapshot-creation-error" description:""`
+	WaitForCreationLockTimeout   encoding.Duration `long:"wait-for-creation-lock-timeout" description:"the maximum a caller to create snapshot should have to wait to acquire the creation lock"`
 }
 
 // NewDefaultConfig creates an instance of the package specific configuration, given a
 // pointer to a logger instance to be used for logging within the package.
 func NewDefaultConfig() Config {
 	return Config{
-		DatabaseSnapshotsCopyToPath:   "/snapshotsCopyTo",
-		DatabaseSnapshotsCopyFromPath: "/snapshotsCopyFrom",
-		WaitForCreationLockTimeout:    encoding.Duration{Duration: 5 * time.Second},
-		PanicOnSnapshotCreationError:  true,
+		WaitForCreationLockTimeout:   encoding.Duration{Duration: 5 * time.Second},
+		PanicOnSnapshotCreationError: true,
 	}
 }

--- a/datanode/dehistory/snapshot/service.go
+++ b/datanode/dehistory/snapshot/service.go
@@ -17,19 +17,22 @@ type Service struct {
 
 	connConfig sqlstore.ConnectionConfig
 
-	createSnapshotLock  mutex.CtxMutex
-	snapshotsCopyToPath string
+	createSnapshotLock    mutex.CtxMutex
+	snapshotsCopyFromPath string
+	snapshotsCopyToPath   string
 }
 
 func NewSnapshotService(log *logging.Logger, config Config, connConfig sqlstore.ConnectionConfig,
+	snapshotsCopyFromPath string,
 	snapshotsCopyToPath string,
 ) (*Service, error) {
 	s := &Service{
-		log:                 log,
-		config:              config,
-		connConfig:          connConfig,
-		createSnapshotLock:  mutex.New(),
-		snapshotsCopyToPath: snapshotsCopyToPath,
+		log:                   log,
+		config:                config,
+		connConfig:            connConfig,
+		createSnapshotLock:    mutex.New(),
+		snapshotsCopyFromPath: snapshotsCopyFromPath,
+		snapshotsCopyToPath:   snapshotsCopyToPath,
 	}
 
 	err := os.MkdirAll(s.snapshotsCopyToPath, fs.ModePerm)

--- a/datanode/dehistory/snapshot/service_create_snapshot.go
+++ b/datanode/dehistory/snapshot/service_create_snapshot.go
@@ -133,8 +133,8 @@ func (b *Service) snapshotData(ctx context.Context, copyDataTx pgx.Tx, dbMetaDat
 
 	start := time.Now()
 	b.log.Infof("copying all table data....")
-	allCopySQL := append(currentSnapshot.GetCopySQL(dbMetaData, b.config.DatabaseSnapshotsCopyToPath),
-		historySnapshot.GetCopySQL(dbMetaData, b.config.DatabaseSnapshotsCopyToPath)...)
+	allCopySQL := append(currentSnapshot.GetCopySQL(dbMetaData, b.snapshotsCopyToPath),
+		historySnapshot.GetCopySQL(dbMetaData, b.snapshotsCopyToPath)...)
 	rowsCopied, err := copyTableData(ctx, copyDataTx, allCopySQL)
 	if err != nil {
 		return fmt.Errorf("failed to copy table data:%w", err)

--- a/datanode/dehistory/snapshot/service_load_snapshot.go
+++ b/datanode/dehistory/snapshot/service_load_snapshot.go
@@ -113,7 +113,7 @@ func (b *Service) loadSnapshot(ctx context.Context, snapshotData compressedFileM
 	b.log.Infof("copying %s into database", snapshotData.UncompressedDataDir())
 
 	rowsCopied, err = b.copyDataIntoDatabase(ctx, vegaDbConn, decompressedFilesDestination,
-		filepath.Join(b.config.DatabaseSnapshotsCopyFromPath, snapshotData.UncompressedDataDir()))
+		filepath.Join(b.snapshotsCopyFromPath, snapshotData.UncompressedDataDir()))
 	if err != nil {
 		b.log.Errorf("failed to copy uncompressed data into the database %s : %w", snapshotData.UncompressedDataDir(), err)
 	}

--- a/datanode/dehistory/store/store.go
+++ b/datanode/dehistory/store/store.go
@@ -317,7 +317,7 @@ func (p *Store) AddSnapshotData(ctx context.Context, historySnapshot snapshot.Hi
 		logging.String("history segment id", contentID.String()),
 		logging.String("chain id", historySegment.ChainID),
 		logging.Int64("from height", historySegment.HeightFrom),
-		logging.Int64("from to", historySegment.HeightTo),
+		logging.Int64("to height", historySegment.HeightTo),
 		logging.String("previous history segment id", historySegment.PreviousHistorySegmentID),
 	)
 

--- a/paths/paths.go
+++ b/paths/paths.go
@@ -275,6 +275,12 @@ var (
 	// DataNodeDeHistoryHome is the folder containing the decentralised history data.
 	DataNodeDeHistoryHome = StatePath(filepath.Join(DataNodeStateHome.String(), "dehistory"))
 
+	// DataNodeDeHistorySnapshotCopyTo is the folder in which the datanode creates snapshots.
+	DataNodeDeHistorySnapshotCopyTo = StatePath(filepath.Join(DataNodeDeHistoryHome.String(), "snapshotscopyto"))
+
+	// DataNodeDeHistorySnapshotCopyFrom is the folder from which the datanode reads snapshot data.
+	DataNodeDeHistorySnapshotCopyFrom = StatePath(filepath.Join(DataNodeDeHistoryHome.String(), "snapshotscopyfrom"))
+
 	// DataNodeEventBufferHome is the folder containing event buffer files.
 	DataNodeEventBufferHome = StatePath(filepath.Join(DataNodeStateHome.String(), "eventsbuffer"))
 

--- a/script/postgres-docker.sh
+++ b/script/postgres-docker.sh
@@ -1,11 +1,11 @@
 #!/bin/bash
 
 if [[ "$OSTYPE" == "linux-gnu"* ]]; then
-        SNAPSHOTS_COPY_TO_PATH=~/.local/state/vega/data-node/dehistory/snapshotsCopyTo
-        SNAPSHOTS_COPY_FROM_PATH=~/.local/state/vega/data-node/dehistory/snapshotsCopyFrom
+        SNAPSHOTS_COPY_TO_PATH=~/.local/state/vega/data-node/dehistory/snapshotscopyto
+        SNAPSHOTS_COPY_FROM_PATH=~/.local/state/vega/data-node/dehistory/snapshotscopyfrom
 elif [[ "$OSTYPE" == "darwin"* ]]; then
-        SNAPSHOTS_COPY_TO_PATH="$HOME/Library/Application Support/vega/data-node/dehistory/snapshotsCopyTo"
-        SNAPSHOTS_COPY_FROM_PATH="$HOME/Library/Application Support/vega/data-node/dehistory/snapshotsCopyFrom"
+        SNAPSHOTS_COPY_TO_PATH="$HOME/Library/Application Support/vega/data-node/dehistory/snapshotscopyto"
+        SNAPSHOTS_COPY_FROM_PATH="$HOME/Library/Application Support/vega/data-node/dehistory/snapshotscopyfrom"
 else
         echo "$OSTYPE" not supported
 fi
@@ -22,6 +22,6 @@ docker run --rm \
            -e POSTGRES_PASSWORD=vega \
            -e POSTGRES_DB=vega \
            -p 5432:5432 \
-           -v "$SNAPSHOTS_COPY_TO_PATH":/snapshotsCopyTo:z \
-           -v "$SNAPSHOTS_COPY_FROM_PATH":/snapshotsCopyFrom:z \
+           -v "$SNAPSHOTS_COPY_TO_PATH":"$SNAPSHOTS_COPY_TO_PATH":z \
+           -v "$SNAPSHOTS_COPY_FROM_PATH":"$SNAPSHOTS_COPY_FROM_PATH":z \
            timescale/timescaledb:2.7.1-pg14

